variantlib/resolver/lib.py

```diff
@@ -0,0 +1,141 @@
+    :param vdescs: list of `VariantDescription` to filter.
+    :param allowed_namespaces: List of allowed variant namespaces as `str`.
+    :param allowed_features: List of allowed `VariantFeature`.
+    :param allowed_properties: List of allowed `VariantProperty`.
+    :return: Filtered list of `VariantDescription`.
+    """
+
+    # ========================= IMPLEMENTATION NOTE ========================= #
+    # Technically, only step 4 is absolutely necessary to filter out the
+    # `VariantDescription` that are not supported on this platform.
+    # 1. There should never be any duplicates on the index
+    #     - filename collision (same filename & same hash)
+    #     - hash collision inside `variants.json`
+    #     => Added for safety and to avoid any potential bugs
+    #     (Note: In all fairness, even if it was to happen, it would most
+    #            likely not be a problem given that we just pick the best match
+    # 2. namespaces are included inside the `VariantProperty` of step 4
+    # 3. features are included inside the `VariantProperty` of step 4
+    #
+    # However, I (Jonathan Dekhtiar) strongly recommend to keep all of them:
+    # - Easier to read and understand
+    # - Easier to debug
+    # - Easier to maintain
+    # - Easier to extend
+    # - Easier to test
+    # - Allows `tooling` to provide CLI/configuration flags to explicitly
+    #   reject a `namespace` or `namespace::feature` without complex
+    #   processing.
+    # ======================================================================= #
+
+    # Step 1: Remove duplicates - should never happen, but just in case
+    result = remove_duplicates(vdescs)
+
+    # Step 2: Remove any `VariantDescription` which declare any `VariantProperty` with
+    # a variant namespace unsupported on this platform.
+    if allowed_namespaces is not None:
+        result = filter_variants_by_namespaces(
+            vdescs=result,
+            allowed_namespaces=allowed_namespaces,
+        )
+
+    # Step 3: Remove any `VariantDescription` which declare any `VariantProperty` with
+    # `namespace :: feature` (aka. `VariantFeature`) unsupported on this platform.
+    if allowed_features is not None:
+        result = filter_variants_by_features(
+            vdescs=result, allowed_features=allowed_features
+        )
+
+    # Step 4: Remove any `VariantDescription` which declare any `VariantProperty`
+    # `namespace :: feature :: value` unsupported on this platform.
+    if allowed_properties is not None:
+        result = filter_variants_by_property(
+            vdescs=result, allowed_properties=allowed_properties
+        )
+
+    yield from result
+
+
+def sort_and_filter_supported_variants(
+    vdescs: list[VariantDescription],
+    supported_vprops: list[VariantProperty] | None = None,
+    property_priorities: list[VariantProperty] | None = None,
+    feature_priorities: list[VariantFeature] | None = None,
+    namespace_priorities: list[str] | None = None,
```

Comment:
```suggestion
    namespace_priorities: list[str] | None = None,
    feature_priorities: list[VariantFeature] | None = None,
    property_priorities: list[VariantProperty] | None = None,
```

Other functions use `(namespace, feature, value)` ordering, so here a similar ordering (with `namespace_priorities` first) seems more consistent.

Reply:
It's fine with me - just would like to mention smthg and confirm you still agree with what you suggest above:

properties > features > namespace>

See `variants.toml`

```toml
# ======================= EXAMPLE `variants.toml` FILE ====================== #

# ============= Top-Most Priority ============= #

# 1. Define the priority of variant properties - 1st order priority.
#    => Expected format: "namespace::feature::value"
#
# OPTIONAL: In most cases - users will not specify the following. Only used when the user wants
#           to modify the default VariantProperty priority ordering.
#
# Note: This is a lazy-list: Only specify the ones you want to "bump up" to the top of the list

property_priority = [
    "fictional_hw::architecture::mother",
    "fictional_tech::risk_exposure:25",
]

# ============= Second-Most Priority ============= #

# 2. Define the priority of variant features - 2nd order priority under the variant properties.
#    => Expected format: "namespace::feature"
#
# OPTIONAL: In most cases - users will not specify the following. Only used when the user wants
#           to modify the default VariantFeature priority ordering.
#
# Note: This is a lazy-list: Only specify the ones you want to "bump up" to the top of the list

features_priority = [
    "fictional_hw::architecture",
    "fictional_tech::risk_exposure",
    "simd_x86_64::feature3",
]

# ============= Default Priority Ordering ============= #

# 3. Define the priority of variant namespaces
#    => Expected format: "namespace"
#
# MANDATORY AND IMPORTANT - PLEASE READ !
#    - As long as there is more than 1 variant plugin installed. This field must be specified.
#      => no default ordering is assumed.
#
#    - If len(plugins) > 1 and an installed plugin is missing in the priority list => raise ConfigurationError
#
#    - If namespaces is specified by no plugin uses it => warning issued.

namespaces_priority = [
    "fictional_hw",
    "fictional_tech",
    "simd_x86_64",
    "non_existent_provider",
]
```

Do you still want to keep the ordering of arguments `(namespace, features, properties)` despite that in this case the "top priority" is property not namespace ?

Reply:
Yeah, I think consistency in argument order (even if they are not strictly the same arguments) is more important — at least I would expect to pass namespace first. You could also look at it as passed in order from lowest priority — first we sort by namespace priorities, then by feature priorities, then by property priorities.

Reply:
That works for me. Changing